### PR TITLE
planner: support hypo tiflash replicas

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #43817

Problem Summary: planner: support hypo tiflash replicas

### What is changed and how it works?

Support `alter table t set hypo tiflash replica {num}`:

```
mysql> create table t (a int);
Query OK, 0 rows affected (0.02 sec)

mysql> explain select a from t;
+-----------------------+----------+-----------+---------------+--------------------------------+
| id                    | estRows  | task      | access object | operator info                  |
+-----------------------+----------+-----------+---------------+--------------------------------+
| TableReader_5         | 10000.00 | root      |               | data:TableFullScan_4           |
| └─TableFullScan_4     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+-----------------------+----------+-----------+---------------+--------------------------------+
2 rows in set (0.00 sec)

mysql> alter table t set hypo tiflash replica 1; -- add a hypo replica
Query OK, 0 rows affected (0.00 sec)

mysql> explain select a from t;
+-------------------------+----------+--------------+---------------+---------------------------------------+
| id                      | estRows  | task         | access object | operator info                         |
+-------------------------+----------+--------------+---------------+---------------------------------------+
| TableReader_11          | 10000.00 | root         |               | MppVersion: 1, data:ExchangeSender_10 |
| └─ExchangeSender_10     | 10000.00 | mpp[tiflash] |               | ExchangeType: PassThrough             |
|   └─TableFullScan_9     | 10000.00 | mpp[tiflash] | table:t       | keep order:false, stats:pseudo        |
+-------------------------+----------+--------------+---------------+---------------------------------------+
3 rows in set (0.00 sec)

mysql> alter table t set hypo tiflash replica 0; -- delete the hypo replica
Query OK, 0 rows affected (0.00 sec)

mysql> explain select a from t;
+-----------------------+----------+-----------+---------------+--------------------------------+
| id                    | estRows  | task      | access object | operator info                  |
+-----------------------+----------+-----------+---------------+--------------------------------+
| TableReader_5         | 10000.00 | root      |               | data:TableFullScan_4           |
| └─TableFullScan_4     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo |
+-----------------------+----------+-----------+---------------+--------------------------------+
2 rows in set (0.00 sec)
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by time-and-fate — APPROVED** [review]

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -1385,9 +1396,10 @@ func getPossibleAccessPaths(ctx sessionctx.Context, tableHints *tableHintInfo, i
 
 	// consider hypo-indexes
 	hypoIndexes := ctx.GetSessionVars().HypoIndexes
-	if hypoIndexes != nil {
-		if hypoIndexes[dbName.L] != nil && hypoIndexes[dbName.L][tblName.L] != nil {
-			for _, index := range hypoIndexes[dbName.L][tblName.L] {
+	if ctx.GetSessionVars().StmtCtx.InExplainStmt && hypoIndexes != nil {
```
Comment (AilinKid) [inline planner/core/planbuilder.go:1399]: why limit it into explain statement？

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

File: planner/core/planbuilder.go (hunk 1 above)
Comment (qw4990) [reply planner/core/planbuilder.go:1399]: Just for safety.

**Review by hawkingrei — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**